bin/runtest.ml

```diff
@@ -33,6 +33,20 @@ let runtest_info =
 ;;
 
 let find_cram_test path ~parent_dir =
+  (* We correct the [path] and [parent_dir] if we have a directory test and the
+     user has given a path to "run.t" rather than the directory cram test itself. *)
+  let path, parent_dir =
+    Option.value ~default:(path, parent_dir)
+    @@
+    let open Option.O in
+    let* basename = Path.Source.basename_opt path in
+    if String.equal basename Source.Cram_test.fname_in_dir_test
```

Comment:
This looks fishy to me, there should be a way to look up tests without having to replicate dune's own logic for discovering them. The logic should be roughly like this:

1. Given a path `foo`, look up the source directory for the parent of this path. Let's call it `parent foo`.
2. If `parent foo` contains a cram test with `chop_prefix (basename foo) ~suffix:".t"`, then we found our test.
3. Otherwise, we check if `foo` is the `run.t` inside a cram test. For that, we now look for a test `parent foo`.
4. if we find such a test, we return it if the "script" of the said test is equal to the user provided path.

All of this logic should hopefully be shared as much as possible with the logic for discovering cram tests in the rules.

Reply:
I pushed out a commit sketching out what I had in mind. Seems to be working